crypto/modes/asm/aes-gcm-avx512.pl

```diff
@@ -0,0 +1,5141 @@
+  my $MASKREG    = $_[7];     # [clobbered] mask register
+  my $CUR_COUNT  = $_[8];     # [out] XMM with current counter (xmm2)
+  my $ZT0        = $_[9];     # [clobbered] ZMM register
+  my $ZT1        = $_[10];    # [clobbered] ZMM register
+  my $ZT2        = $_[11];    # [clobbered] ZMM register
+  my $ZT3        = $_[12];    # [clobbered] ZMM register
+  my $ZT4        = $_[13];    # [clobbered] ZMM register
+  my $ZT5        = $_[14];    # [clobbered] ZMM register
+  my $ZT6        = $_[15];    # [clobbered] ZMM register
+  my $ZT7        = $_[16];    # [clobbered] ZMM register
+  my $ZT8        = $_[17];    # [clobbered] ZMM register
+  my $ZT9        = $_[18];    # [clobbered] ZMM register
+  my $ZT10       = $_[19];    # [clobbered] ZMM register
+  my $ZT11       = $_[20];    # [clobbered] ZMM register
+  my $ZT12       = $_[21];    # [clobbered] ZMM register
+  my $ZT13       = $_[22];    # [clobbered] ZMM register
+  my $ZT14       = $_[23];    # [clobbered] ZMM register
+  my $ZT15       = $_[24];    # [clobbered] ZMM register
+  my $ZT16       = $_[25];    # [clobbered] ZMM register
+  my $ZT17       = $_[26];    # [clobbered] ZMM register
+
+  my $ZT0x = $ZT0;
+  $ZT0x =~ s/zmm/xmm/;
+
+  $code .= <<___;
+        cmp     \$12,$IV_LEN
+        je      iv_len_12_init_IV
+___
+
+  # ;; IV is different than 12 bytes
+  &CALC_J0($GCM128_CTX, $IV, $IV_LEN, $CUR_COUNT, $ZT0, $ZT1, $ZT2, $ZT3, $ZT4, $ZT5, $ZT6, $ZT7,
+    $ZT8, $ZT9, $ZT10, $ZT11, $ZT12, $ZT13, $ZT14, $ZT15, $ZT16, $ZT17, $GPR1, $GPR2, $GPR3, $MASKREG);
+  $code .= <<___;
+       jmp      skip_iv_len_12_init_IV
+iv_len_12_init_IV:   # ;; IV is 12 bytes
+        # ;; read 12 IV bytes and pad with 0x00000001
+        vmovdqu8          ONEf(%rip),$CUR_COUNT
+        mov               $IV,$GPR2
+        mov               \$0x0000000000000fff,@{[DWORD($GPR1)]}
+        kmovq             $GPR1,$MASKREG
+        vmovdqu8          ($GPR2),${CUR_COUNT}{$MASKREG}         # ; ctr = IV | 0x1
+skip_iv_len_12_init_IV:
+        vmovdqu           $CUR_COUNT,$ZT0x
+___
+  &ENCRYPT_SINGLE_BLOCK($AES_KEYS, "$ZT0x", "$GPR1");    # ; E(K, Y0)
+  $code .= <<___;
+        vmovdqu           $ZT0x,`$CTX_OFFSET_EK0`($GCM128_CTX)   # ; save EK0 for finalization stage
+
+        # ;; store IV as counter in LE format
+        vpshufb           SHUF_MASK(%rip),$CUR_COUNT,$CUR_COUNT
+        vmovdqu           $CUR_COUNT,`$CTX_OFFSET_CurCount`($GCM128_CTX)   # ; save current counter Yi
+___
+}
+
+sub GCM_UPDATE_AAD {
+  my $GCM128_CTX = $_[0];     # [in] GCM context pointer
+  my $A_IN       = $_[1];     # [in] AAD pointer
+  my $A_LEN      = $_[2];     # [in] AAD length in bytes
+  my $GPR1       = $_[3];     # [clobbered] GP register
+  my $GPR2       = $_[4];     # [clobbered] GP register
+  my $GPR3       = $_[5];     # [clobbered] GP register
+  my $MASKREG    = $_[6];     # [clobbered] mask register
+  my $AAD_HASH   = $_[7];     # [out] XMM for AAD_HASH value (xmm14)
```

Comment:
should there be an assumption about the exact xmm register used here? i.e. any xmm could potentially be used for $AAD_HASH by a caller of the GCM_UPDATE_AAD subroutine? (not just xmm14)

Reply:
Agree, removed.